### ISSUE TITLE
Re-exported modules from index.ts are not getting tree shaken

### ISSUE THREAD

**Issue by Poolshark** [vercel/next.js#67584]
### Link to the code that reproduces this issue

https://github.com/Poolshark/nextjs-bugreport

### To Reproduce

1. Run application with `pnpm dev`
2. Verify that the application is running → it should display "Hello World!"
3. Open `middlewares/index.ts` and un-comment line 2
4. Refresh the browser window and verify that the application now throws an error

###### Current
Currently, we re-export two modules from `middlewares/index.ts`. We only use one module in `middleware.ts`. The other module has a `nodemailer` dependency which, when used in a middleware, results in the error
```
Error: The edge runtime does not support Node.js 'stream' module.
Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime
```
Although we do not use the module with the `nodemailer` dependency in the middleware, we get the error which clearly indicates that it gets bundled with the source.

###### Expected
We would expect that, since we do not use module 2 in the middleware, the source is not part of the bundle and thus no error is thrown even if the re-export statement in `index.ts` is present.

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000
  Available memory (MB): 32768
  Available CPU cores: 10
Binaries:
  Node: 22.4.0
  npm: 10.8.1
  Yarn: 1.22.22
  pnpm: 9.4.0
Relevant Packages:
  next: 15.0.0-canary.58 // There is a newer canary version (15.0.0-canary.59) available, please upgrade! 
  eslint-config-next: N/A
  react: 19.0.0-rc-6f23540c7d-20240528
  react-dom: 19.0.0-rc-6f23540c7d-20240528
  typescript: 5.3.3
Next.js Config:
  output: N/A
 ⚠ There is a newer canary version (15.0.0-canary.59) available, please upgrade! 
   Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
   Read more - https://nextjs.org/docs/messages/opening-an-issue
```

### Which area(s) are affected? (Select all that apply)

Not sure, Developer Experience, Webpack

### Which stage(s) are affected? (Select all that apply)

next dev (local)

### Additional context

I have only tested this with the local dev server but not with a production build. However, even in the case this issue does not appear in a production build it would be badly impacting the developer experience.

**Comment by SlimDaddy228** [comment]
I have the same problem, only related to the fact that it exports all the constants from one file to all pages in a whole bunch

If you need code that reproduces this, or help, please write, the solution is very necessary!

**Comment by MrOxMasTer** [comment]
I do not know if this is related to this problem, but I have been observing this for a long time.

The problem is the re-export of client and server files that cannot live together. An error is occurring.

For example, when I try to make a single business entity, but my api is server-side, my models schemes can be used both on the server and on the client, and the UI is also server/client-side.

How to combine this into one index.ts is unknown to me, since an error appears that, for example, either the server function is not performed on the server or, conversely, "use client" is required

**Comment by Poolshark** [comment]
Although it took me a while to get back to this, I finally found a solution. I have had a [similar issue](https://github.com/webpack/webpack/discussions/15854) with a mono repo a while ago and the problem was related to setting the `sideEffects` flag correctly.

Since Next.js is using webpack under the hood, proper tree shaking can only be achieved when setting `sideEffects` correctly. I could solve my bundling issue with simply setting

```json
{
  "sideEffects": false,
}
```

in my `package.json`. Since I did not have to set `optimization.sideEffects: true` in the Webpack config, I assume that this is already the default setting from Next.js.

Hope this helps! 🙌
